### ISSUE TITLE
False positive regression for `unused_crate_dependencies`

### ISSUE THREAD

**Issue by robertbastian** [rust-lang/rust#147069]
In beta (1.91.0-beta.3 (24bb93c38 2025-09-10)), there is a false positive from the `unused_crate_dependencies` lint that does not happen on 1.90.

To reproduce:
* clone https://github.com/unicode-org/icu4x at `292384d6a04746164faee7d660e75a5c7ca2b89c`
* `RUSTFLAGS="-D unused_crate_dependencies" cargo check --all-features --manifest-path components/properties/Cargo.toml`
  * ✅
* `RUSTFLAGS="-D unused_crate_dependencies" cargo +beta check --all-features --manifest-path components/properties/Cargo.toml`

```console
error: extern crate `icu_locale_core` is unused in crate `icu_properties`
 --> components/properties/src/lib.rs:5:1
  |
5 | //! Definitions of [Unicode Properties] and APIs for
  | ^
  |
  = help: remove the dependency or add `use icu_locale_core as _;` to the crate root
  = note: requested on the command line with `-D unused-crate-dependencies`

error: could not compile `icu_properties` (lib) due to 1 previous error
```

This crate is used, so it's regression, not a bug fix.

**Comment by lukaslueg** [comment]
Bisects to #144793, cc @petrochenkov

searched nightlies: from nightly-2025-06-23 to nightly-2025-09-26
regressed nightly: nightly-2025-08-15
searched commit range: https://github.com/rust-lang/rust/compare/3672a55b7cfd0a12e7097197b6242872473ffaa7...898aff704d6f0d00343f21d31b8b9bfac8e43007
regressed commit: https://github.com/rust-lang/rust/commit/8e7795415acaa9471b5eeb2301915d6996d289ec

**Comment by apiraino** [comment]
Assigning priority (discussion on [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/245100-t-compiler.2Fprioritization.2Falerts/topic/.23147069.20False.20positive.20regression.20for.20.60unused_crate_depende.E2.80.A6/near/543161973)).

@rustbot label -I-prioritize +P-medium

**Comment by epage** [comment]
As part of https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#lintscargo, Cargo has a lint that wraps `unused_crate_dependencies`, see https://doc.rust-lang.org/nightly/cargo/reference/lints.html#unused_dependencies

If this is really for a feature activation, that is something we are still investigating how we want to handle.  We have a `lints.cargo.unused_dependencies.ignore` field but I'm exploring alternatives in https://github.com/rust-lang/rfcs/pull/3920.
